include/ldsCtrlEst_h/lds_mpcctrl.h

```diff
@@ -0,0 +1,533 @@
+   */
+  MpcController() = default;
+
+  /**
+   * @brief     Constructs a new MpcController.
+   *
+   * @param     sys   The system being controlled
+   * @param     u_lb  The lower bound of the control input
+   * @param     u_ub  The upper bound of the control input
+   *
+   * @tparam    System  The system type
+   */
+  MpcController(const System& sys, Vector u_lb, Vector u_ub);
+
+  /**
+   * @brief     Constructs a new MpcController.
+   *
+   * @param     sys   The system being controlled
+   * @param     u_lb  The lower bound of the control input
+   * @param     u_ub  The upper bound of the control input
+   *
+   * @tparam    System  The system type
+   */
+  MpcController(System&& sys, Vector u_lb, Vector u_ub);
+
+  /**
+   *
+   */
+  Vector Control(data_t t_sim, const Vector& x0, const Vector& u0,
+                 const Matrix& xr);
+
+  // getters
+
+  // setters
+  void set_control(Matrix Q, Matrix R, Matrix S, size_t N, size_t M) {
+    Q_ = Q;
+    // R_ = R; // Isn't used
+    S_ = S;
+    N_ = N;
+    M_ = M;
+
+    // Set up P matrix
+    Matrix Px   = arma::kron(Matrix(N_, N_, arma::fill::eye), Q_);
+    Matrix Pu1  = arma::kron(Matrix(M_, M_, arma::fill::eye), 2 * S_ + R);
+    Matrix Pu2  = arma::kron(Matrix(eye_offset(M)) + Matrix(eye_offset(M, 1)), -S_);
+    Matrix Pu3  = block_diag(Matrix((M_ - 1) * m_, (M_ - 1) * m_, arma::fill::zeros), -S_);
+    Matrix Pu = Pu1 + Pu2 + Pu3;
+    P_ = arma::trimatu(2 * block_diag(Px, Pu)); // Taking only the upper triangular part
+  }
+
+  void set_constraint(Vector xmin, Vector xmax, Vector umin, Vector umax) {
+    lineq_ = join_horiz(arma::kron(Vector(N_, arma::fill::ones), xmin).t(),
+                        arma::kron(Vector(M_, arma::fill::ones), umin).t());
+    uineq_ = join_horiz(arma::kron(Vector(N_, arma::fill::ones), xmax).t(),
+                        arma::kron(Vector(M_, arma::fill::ones), umax).t());
+    size_t Aineq_dim = N_ * n_ + M_ * m_;
+    Aineq_ = arma::eye<arma::SpMat<data_t>>(Aineq_dim, Aineq_dim);
+  }
+
+  void Print() {
+    sys_.Print();
+    // TODO: Implement print
+  }
+
+ protected:
+  System sys_;  ///< system being controlled
+  size_t n_;    ///< number of states
+  size_t m_;    ///< number of output states
+  size_t N_;    ///< number of steps
+  size_t M_;    ///< number of inputs
+  Matrix A_;    ///< state transition matrix
+  Matrix B_;    ///< input matrix
+  Matrix P_;    ///< penalty matrix
+  Matrix Q_;
+  Matrix S_;
+
+  Matrix lineq_;               ///< lower inequality bound
+  Matrix uineq_;               ///< upper inequality bound
+  arma::SpMat<data_t> Aineq_;  ///< inequality condition matrix
+
+  Matrix Acon_;  ///< update condition matrix
+  Vector lb_;    ///< lower bound
+  Vector ub_;    ///< upper bound
+
+  Vector xi_;     ///< previous step end state
+  size_t t_sim_;  ///< previous step simulation time step
+
+ private:
+  /**
+   * @brief     Set the matrix for the OSQP solver from an Armadillo matrix
```

Comment:
this doxygen comment is misleading; it sounds like it will directly set a matrix on the controller, but instead just does the conversion

Reply:
and for from_sparse as well

Reply:
Ok, I will fix that!